File: 2. Add Two Numbers/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = new ListNode();
        ListNode current = head;
        int carry = 0;

        while (l1 != null || l2 != null || carry != 0) {
            current.next = new ListNode();
            current = current.next;

            int l1val = 0;
            if (l1 != null) {
                l1val = l1.val;
                l1 = l1.next;
            }

            int l2val = 0;
            if (l2 != null) {
                l2val = l2.val;
                l2 = l2.next;
            }

            int sum = carry + l1val + l2val;
            if (sum < 10) {
                carry = 0;
            } else {
                carry = 1;
                sum -= 10;
            }

            current.val = sum;
        }

        return head.next;
    }
}

File: 2. Add Two Numbers/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function addTwoNumbers(
    l1: ListNode | null,
    l2: ListNode | null
): ListNode | null {
    const head = new ListNode();
    let current = head;
    let carry = 0;

    while (l1 || l2 || carry) {
        current.next = new ListNode();
        current = current.next;
        let sum = carry + (l1?.val ?? 0) + (l2?.val ?? 0);

        if (sum < 10) {
            carry = 0;
        } else {
            carry = 1;
            sum -= 10;
        }

        current.val = sum;
        l1 = l1?.next;
        l2 = l2?.next;
    }

    return head.next;
}
